### PR TITLE
Make extractors, predictors, and TrainPreprocessor configurable in string format.

### LINKED ISSUE #413: Extractors, Predictors, and TrainPreprocesor cannot be configured in string format.
**Is your feature request related to a problem? Please describe.**
When building training pipeline, I want to make components can be configured in string format as much as possible, so that users would only need to modify yaml config files to build training pipelines with minimum touch of the python scripts. 
For example, in [this pull request](https://github.com/asyml/forte/pull/318/files), users have to modify tagging_trainer.py a lot to build their training pipeline, which is due to the configuration of extractors, predictors, and TrainPreprocessors require instantiated class as input.

**Describe the solution you'd like**
Change configuration method so that it only need the full path (string) of a class then instantiate it before using. 

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by jieralice13**
This PR fixes #413, #414, #415 and #417.

### Description of changes
Modified configuration method of all extractors, predictor, and TrainPreprocessor to make them configurable in string format.
Modified Vocabulary configurable variables to include pad_value and unk_value.
Modified batch generation loop in batcher.py to put it into a list, then iterate over it.
Added do_eval to Predictor to make the user choose do or not do pre_evaluation_action.

### Possible influences of this PR.
Describe what are the possible side-effects of the code change.

### Test Conducted
A reuse-processor test is added. The testing processor will record the number of times it gets initialized. And the number should be 1.

### CODE REVIEW

**Comment by hunterhector** [comment]
Can you fix the pytest errors?

**Comment by hunterhector** [comment]
Btw, for the PR, please use our PR template. And also create an issue for this PR to solve.

File: forte/data/extractors/subword_extractor.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+        """
+        data = []
+        for subword in pack.get(self._entry_type, instance):
+            text = subword.text
+            if not subword.is_first_segment:
+                text = '##' + text
+            if text in self.tokenizer.vocab:
+                data.append(self.tokenizer.vocab[text])
+            else:
+                data.append(self.tokenizer.vocab['[UNK]'])
```
Comment (hunterhector) [inline forte/data/extractors/subword_extractor.py:87]: I tried to unify the special tokens with Texar-pytorch, in this PR, you can check: https://github.com/asyml/forte/pull/409/files#diff-79b7601c2b37366060c6d52de883a3ad381a765c84b0f54f6276a18b6e17530cR122

File: forte/processors/base/batch_processor.py [hunk 2]
```diff
@@ -288,24 +292,61 @@ def default_configs(cls) -> Dict[str, Any]:
...
+                config=scheme["extractor"]["config"])
+            if "vocab_path" in scheme["extractor"]:
+                vocab_file = open(scheme["extractor"]["vocab_path"], 'rb')
+                extractor.vocab = pickle.load(vocab_file)
+                vocab_file.close()
+            parsed_configs["feature_scheme"][tag]["extractor"] = \
+                extractor
+
+            if "converter" not in scheme:
+                parsed_configs["feature_scheme"][tag]["converter"] = \
```
Comment (hunterhector) [inline forte/processors/base/batch_processor.py:343]: So we are storing the objects inside the `Config`, can we just store them in some other dict? We probably don't want to demo that configs can have objects/instances.

Fore example, we just store something named `self.feature_scheme`?

**Review by hunterhector — COMMENTED** [review]
Thanks for the hard work! There are a few suggestions inline. And we need to solve more CI issues.

File: forte/data/base_extractor.py [hunk 3]
```diff
@@ -125,21 +134,35 @@ def default_configs(cls):
         "vocab_use_unk" (bool)
             Whether the `<UNK>` element should be added to vocabulary.
             Default is true.
+
+        "pad_value" (int)
+            ID assigned to pad. It should be integer smaller than 0.
```
[the remark this answers is not in the document]
Comment (jieralice13) [reply forte/data/base_extractor.py:139]: When we need the pad_value to be a specific one, such as -100

[the remark this answers is not in the document]
Comment (hunterhector) [reply forte/data/base_extractor.py:139]: That's the `representation` of the vocab, not the id? See here:

https://github.com/asyml/forte/pull/409/files#diff-79b7601c2b37366060c6d52de883a3ad381a765c84b0f54f6276a18b6e17530cR203

and here:

https://github.com/asyml/forte/pull/409/files#diff-79b7601c2b37366060c6d52de883a3ad381a765c84b0f54f6276a18b6e17530cR253

File: forte/data/batchers.py [hunk 4]
```diff
@@ -274,7 +274,7 @@ def _get_data_batch(
         features_collection: List[Dict[str, Feature]] = []
         current_size = self.pool_size
 
-        for instance in data_pack.get(self.scope):
+        for instance in list(data_pack.get(self.scope)):
```
[the remark this answers is not in the document]
Comment (jieralice13) [reply forte/data/batchers.py:277]: Please see this bug report #415

File: forte/data/extractors/seqtagging_extractor.py [hunk 5]
```diff
@@ -46,51 +47,78 @@ class BioSeqTaggingExtractor(BaseExtractor):
...
         if self.config.attribute is None:
             raise AttributeError("attribute is required "
                                  "in BioSeqTaggingExtractor.")
-        if self.config.tagging_unit is None:
+        if self.config.tagging_unit == "":
             raise AttributeError("tagging_unit is required in "
                                  "BioSeqTaggingExtractor.")
+        self.attribute = self.config.attribute
+        self.tagging_unit = get_class(self.config.tagging_unit)
+        self.is_bert = self.config.is_bert
```
[the remark this answers is not in the document]
Comment (jieralice13) [reply forte/data/extractors/seqtagging_extractor.py:61]: It's not only related to subwords. It's related to Bert specific action. As mentioned in the docstring below, "It indicates whether Bert model is used. If true, padding will be added to the begining and end of a sentence corresponding to the special tokens ([CLS], [SEP]) used in Bert."

File: forte/data/batchers.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (hunterhector) [reply forte/data/batchers.py:277]: Do you know why this happens and why the solution works?

File: forte/data/base_extractor.py [hunk 6]
```diff
@@ -125,21 +134,35 @@ def default_configs(cls):
...
         }
 
     @property
-    def entry_type(self) -> Type[Annotation]:
-        return self.config.entry_type
+    def entry_type(self) -> object:
+        return self._entry_type
+
+    @entry_type.setter
+    def entry_type(self, input_entry: Annotation):
```
[the remark this answers is not in the document]
Comment (hunterhector) [reply forte/data/base_extractor.py:160]: Can you comment the resolution before mark this as resolved?

**Comment by hunterhector** [comment]
Let's keep the conversation open before we really resolve them, either via a new commit, or a discussion.

**Review by hunterhector — APPROVED** [review]
Most should be fine, there are a few small details but I am not going to dig into those in this PR.

**Pull request merged**